### DISCUSSION TITLE
Question about the arguments type and the return type in specific methods

### DISCUSSION THREAD

**Discussion by shikouzhong** [mathnet/mathnet-numerics#813] [Q&A] [answered]
C# is a strongly-typed language, when I use MathNet.Numerics in C#, this put me to trouble.

for example:
when I call the `PointwiseMultiply()` method, the former matrix and the latter one must be the same type. That means I must convert a `Matrix<float>` into a `Matrix<Complex32>` , when it pointwise multiplies with a `Matrix<Complex32>` of the same dimension. 

But we don't need to do that in MATLAB, so I'm wondering is there any better practice?
Excuse me for my simple question. :-)

**Comment by cdrnet** [comment, marked as answer]
The type-safety is intentional and generally desired and wanted. It does mean that you can, but also have to, be aware of what domain you're computing in.

Technically we could offer all operations also cross-type, unfortunately this would explode the API in a way that is hard to justify. So you'll essentially have to find a good boundary in your algorithms between real and complex domain areas, and convert whenever you have to cross that boundary, e.g. using `ToComplex32()`.

**Comment by shikouzhong** [comment]
I got it. Thanks
